kratos/spatial_containers/spatial_search_result_container.cpp

```diff
@@ -0,0 +1,481 @@
+            }
+        } else {
+            // Sending local results if any
+            if (local_result_size > 0) {
+                std::vector<GlobalPointerResultType> local_gp;
+                local_gp.reserve(local_result_size);
+                for (auto& r_value : mLocalResults) {
+                    local_gp.push_back(GlobalPointerResultType(&r_value, rank));
+                }
+                rDataCommunicator.Send(local_gp, 0);
+            }
+
+            // Receiving synced result
+            std::vector<GlobalPointerResultType> global_gp;
+            rDataCommunicator.Recv(global_gp, 0);
+
+            // Transfer to global pointer
+            for (auto& r_gp : global_gp) {
+                mGlobalResults.push_back(r_gp);
+            }
+        }
+    } else { // Serial code
+        // Fill global vector
+        for (auto& r_value : mLocalResults) {
+            auto p_result = Kratos::make_shared<SpatialSearchResultType>(r_value);
+            mGlobalResults.push_back(p_result);
+        }
+    }
+
+    // Generate the communicator
+    mpGlobalPointerCommunicator = Kratos::make_shared<GlobalPointerCommunicatorType>(rDataCommunicator, mGlobalResults.ptr_begin(), mGlobalResults.ptr_end());
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template <class TObjectType>
+std::vector<double> SpatialSearchResultContainer<TObjectType>::GetDistances()
+{
+    // Check if the communicator has been created
+    KRATOS_ERROR_IF(mpGlobalPointerCommunicator == nullptr) << "The communicator has not been created." << std::endl;
+    
+    // Define the coordinates vector
+    const std::size_t number_of_gp = mGlobalResults.size();
+    std::vector<double> distances(number_of_gp);
+
+    // Call Apply to get the proxy
+    auto proxy = this->Apply([](GlobalPointerResultType& rGP) -> double {
+        return rGP->GetDistance();
+    });
+
+    // Get the coordinates
+    for(std::size_t i=0; i<number_of_gp; ++i) {
+        auto& r_gp = mGlobalResults(i);
+        distances[i] = proxy.Get(r_gp);
+    }
+
+    return distances;
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template <class TObjectType>
+std::vector<Vector> SpatialSearchResultContainer<TObjectType>::GetResultShapeFunctions(const array_1d<double, 3>& rPoint)
+{
+    // Check if the communicator has been created
+    KRATOS_ERROR_IF(mpGlobalPointerCommunicator == nullptr) << "The communicator has not been created." << std::endl;
+
+    // Define the coordinates vector
+    const std::size_t number_of_gp = mGlobalResults.size();
+    std::vector<Vector> shape_functions(number_of_gp);
+
+    // Call Apply to get the proxy
+    auto proxy = this->Apply([&rPoint](GlobalPointerResultType& rGP) -> Vector {
```

Comment:
this is a blocking call, it shouldn't be called for a single point

Reply:
Let me see

Reply:
This calculates all shape functions in a given point, so it is computing in multiple "resulting points" only one given point. Read the code again and you will see it.

Reply:
i think we need to have a discussion in order to clarify this.

Reply:
@jcotela we mean this. I am using global pointer communicator as I assume it is designed